File: bindings/prql-php/tests/CompilerTest.php
<?php

declare(strict_types=1);

namespace Prql\Compiler\Test;

use Prql\Compiler\Compiler;
use Prql\Compiler\Options;
use PHPUnit\Framework\TestCase;

final class CompilerTest extends TestCase
{
    public function testFfiExtensionIsLoaded(): void
    {
        $this->assertTrue(extension_loaded("ffi"));
    }

    public function testPrqlLibraryFileExists(): void
    {
        $this->assertFileExists("lib/libprql_lib.so");
    }

    public function testPrqlHeaderFileExists(): void
    {
        $this->assertFileExists("lib/libprql_lib.h");
    }

    public function testInvalidQuery(): void
    {
        $prql = new Compiler();
        $res = $prql->compile("invalid");

        $this->assertCount(1, $res->messages);
    }

    public function testCompileWorks(): void
    {
        $options = new Options();
        $options->format = false;
        $options->signature_comment = false;
        $options->target = "sql.mssql";
        $prql = new Compiler();

        $actual = $prql->compile("from employees | take 10", $options);
        $this->assertCount(0, $actual->messages);

        $this->assertEquals("SELECT TOP (10) * FROM employees", $actual->output);
    }

    public function testOtherFunctions(): void
    {
        $prql = new Compiler();

        $query = "
            let a = (from employees | take 10)

            from a | select {first_name}
        ";

        $pl = $prql->prqlToPL($query);
        $this->assertCount(0, $pl->messages);

        $rq = $prql->plToRQ($pl->output);
        $this->assertCount(0, $rq->messages);

        $via_json = $prql->rqToSQL($rq->output);
        $this->assertCount(0, $via_json->messages);

        $direct = $prql->compile($query);
        $this->assertCount(0, $direct->messages);

        $this->assertEquals($via_json, $direct);
    }
}

File: prqlc/bindings/dotnet/PrqlCompiler.Tests/CompilerTest.cs
using Prql.Compiler;

namespace Prql.Compiler.Tests;

sealed public class CompilerTest
{
    [Fact]
    public void ToCompile_Works()
    {
    	// Arrange
        var expected = "SELECT * FROM employees";
        var options = new PrqlCompilerOptions
        {
            Format = false,
            SignatureComment = false,
            Target = "sql.mssql"
        };

        // Act
        var result = PrqlCompiler.Compile("from employees", options);

        // Assert
        Assert.Equal(expected, result.Output);
    }

    [Fact]
    public void Compile_ReportsErrorMessages()
    {
        // Arrange — `unknown_function` is not defined, producing an error
        // message whose optional fields (Span, Display, Location) get
        // populated. This validates pointer dereferencing for indirect
        // string fields and Span/Location pointers in the FFI struct layout.
        var query = "from employees | unknown_function col";

        // Act
        var result = PrqlCompiler.Compile(query);

        // Assert
        Assert.NotEmpty(result.Messages);
        var message = result.Messages.First();
        Assert.Equal(MessageKind.Error, message.Kind);
        Assert.False(string.IsNullOrEmpty(message.Reason));
        Assert.NotNull(message.Span);
        Assert.NotNull(message.Location);
        Assert.False(string.IsNullOrEmpty(message.Display));
    }

    [Fact]
    public void Compile_ThrowsArgumentNullException_WhenOptionsNull()
    {
        Assert.Throws<ArgumentNullException>(() => PrqlCompiler.Compile("from x", null!));
    }

    [Fact]
    public void RqToSql_ThrowsArgumentNullException_WhenOptionsNull()
    {
        Assert.Throws<ArgumentNullException>(() => PrqlCompiler.RqToSql("{}", null!));
    }

    [Fact]
    public void TestOtherFunctions()
    {
        // Arrange
        var query = """
            let a = (from employees | take 10)

            from a | select {first_name}
            """;
        var options = new PrqlCompilerOptions();

        // Act and assert
        var pl = PrqlCompiler.PrqlToPl(query);
        Assert.Empty(pl.Messages);

        var rq = PrqlCompiler.PlToRq(pl.Output);
        Assert.Empty(rq.Messages);

        var via_json = PrqlCompiler.RqToSql(rq.Output, options);
        Assert.Empty(via_json.Messages);

        var direct = PrqlCompiler.Compile(query);
        Assert.Empty(direct.Messages);

        Assert.Equal(via_json.Messages, direct.Messages);
        Assert.Equal(via_json.Output, direct.Output);
    }
}
